deepinv/physics/phase_retrieval.py

```diff
@@ -0,0 +1,127 @@
+from deepinv.physics.forward import Physics, LinearPhysics
+from deepinv.physics.compressed_sensing import CompressedSensing
+import torch
+import numpy as np
+
+
+class PhaseRetrieval(Physics):
+    r"""
+    Phase Retrieval forward operator. The linear operator is defined by a :meth:`deepinv.physics.LinearPhysics` object with :math:`B`.
+
+    An existing operator can be loaded from a saved .pth file via ``self.load_state_dict(save_path)``, in a similar fashion to :class:`torch.nn.Module`.
+
+    :param deepinv.physics.forward.LinearPhysics B: the linear forward operator.
+    """
+
+    def __init__(
+        self,
+        B: LinearPhysics,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.name = f"PR_m{self.m}"
+
+        self.B = B
+
+    def A(self, x: torch.Tensor) -> torch.Tensor:
+        r"""
+        Applies the forward operator to the input x.
+
+        Note here the operation includes the modulus operation.
+
+        :param torch.Tensor x: signal/image.
+        """
+        return self.B(x).abs().square()
+
+    def A_adjoint(self, y: torch.Tensor) -> torch.Tensor:
+        return self.B.A_adjoint(y)
+
+    def A_dagger(self, y):
+        r"""
+        Computes the linear pseudo-inverse of :math:`B`.
+
+        :param torch.Tensor y: signal/image.
+        :return: (torch.Tensor) the reconstruction image :math:`x`.
+        """
+        return self.B.A_dagger(y)
+
+    def forward(self, x):
+        r"""
+        Applies the phase retrieval measurement operator, i.e. :math:`y = N(|Bx|^2)` (with noise :math:`N` and/or sensor non-linearities).
+
+        :param torch.Tensor,list[torch.Tensor] x: signal/image
+        :return: (torch.Tensor) noisy measurements
+        """
+        return self.sensor(self.noise(self.A(x)))
+
+    def A_jvp(self, x, v):
+        r"""
+        Computes the product between the Jacobian of the forward operator :math:`A` at the input x and perturbation v, defined as:
+
+        .. math::
+
+            A_{jvp}(x, v) = 2 \overline{B}^{\top} diag(Bx) v.
+
+        :param torch.Tensor x: signal/image.
+        :param torch.Tensor v: perturbation.
+        :return: (torch.Tensor) the Jacobian of the forward operator at x in the direction v.
+        """
+        return 2 * self.A_adjoint(self.B(x) * v)
+
+
+class RandomPhaseRetrieval(PhaseRetrieval):
+    r"""
+    Random Phase Retrieval forward operator. Creates a random :math:`m \times n` sampling matrix where :math:`n` is the number of elements of the signal and :math:`m` is the number of measurements.
```

Comment:
sampling matrix :math:`B` (to make it easier to relate to the phase retrieval equation in the base class)